locator/tablets.cc

```diff
@@ -245,9 +245,32 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
     }
 }
 
+const tablet_replica tablet_map::sort_and_get_primary_replica(const tablet_replica_set& candidates, tablet_id id, std::function<bool(const tablet_replica&)> filter_func) {
+    // clone the replica set since we mustn't sort it in-place.
+    auto replicas = filter_func ? boost::copy_range<tablet_replica_set>(candidates | boost::adaptors::filtered(filter_func)) : candidates;
+    size_t size = replicas.size();
+    size_t idx = id.value() % size;
+    // partially sort as few items as possible
+    // although partial_sort doesn't guarantee stable sort of equal items
+    // we compare here host_id:s that must be unique, so there is no problem
+    // using std:;partial_sort
+    if (idx < size / 2) {
+        // sort the first idx replicas in ascending order
+        std::partial_sort(replicas.begin(), replicas.begin() + idx + 1, replicas.end(), [] (const tablet_replica& l, const tablet_replica& r) {
+            return l.host < r.host;
+        });
+    } else {
+        // sort the last (size - idx) replicas in descending order
+        std::partial_sort(replicas.rbegin(), replicas.rbegin() + size - idx, replicas.rend(), [] (const tablet_replica& l, const tablet_replica& r) {
+            return l.host > r.host;
+        });
+    }
```

Comment:
Are you looking for nth_element?

Note std::ranges::partial_sort accepts a projection so you can pass std::mem_fn(&tablet_replica::host).

I'm not convinced this is less vulnerable to bad distributions than the original.

Reply:
> Are you looking for nth_element?

yes

> Note std::ranges::partial_sort accepts a projection so you can pass std::mem_fn(&tablet_replica::host).


That would be std::ranges::partial_sort, but yes, it would simplify this a bit.

> I'm not convinced this is less vulnerable to bad distributions than the original.

We can consider the nodes' network topology so to rotate DCs and RACKs before nodes (in host_id, or node->idx order).

Reply:
> but I'm not sure it's worth it to lookup the node for each host, for each tablet.

Hmm, we do it for get_primary_replica_within_dc anyways...  so maybe it isn't so bad.

Reply:
> 
> > I'm not convinced this is less vulnerable to bad distributions than the original.
> 
> We can consider the nodes' network topology so to rotate DCs and RACKs before nodes (in host_id, or node->idx order).

You're fixing a problem I don't understand. Yes, if the tablet replicas are exactly the wrong way, we get a lopsided primary owner. But don't we also get a lopsided primary owner with any other algorithm, just with a different worst-case arrangement?

Reply:
The problem I was thinking about is distributung thd primary owner work evenly across datacenters and rack, with a small number of tablets.
If there are enough tablets, with the ordinal selection over sorted replicas, by host_id, there shouldn't be any such unfairness.

However, if indeed the tablets are imbalanced and some nodes own considerably more tablets than others, then the bias will be reflected on the primary replicas as well, since on average the number of primary replicas on each node will be the number of tablets on the node divided by the total number of replicas.

Reply:
If we keep the primary assignment in the map, we can balance the selection by keeping a tally of number of primary replicas per node, similar to the number of replicas per node and shard we use for tablet allocation.
But this will need to get updated on tablet transitions.

Reply:
I implemented load-balancing approach in v9
by keeping a tally of number of primary replicas per node
so to distributed the primary-replica load evenly across all tablets in a table.
This is not kept anywhere.  It is valid per the immutable tablet_map (that's held alive with the erm/token_metadata_ptr).